x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ClusterPermission.java

```diff
@@ -5,122 +5,198 @@
+            }
+            return new ClusterPermission(this.clusterPrivileges, checks);
         }
     }
 
     /**
-     * A permission that makes use of both cluster privileges and request inspection
+     * Evaluates whether the cluster actions (optionally for a given request)
+     * is permitted by this permission.
      */
-    public static class ConditionalClusterPermission extends ClusterPermission {
-        private final ConfigurableClusterPrivilege conditionalPrivilege;
+    public interface PermissionCheck {
+        /**
+         * Checks permission to a cluster action for a given request.
+         *
+         * @param action  action name
+         * @param request {@link TransportRequest}
+         * @return {@code true} if the specified action for given request is allowed else returns {@code false}
+         */
+        boolean check(String action, TransportRequest request);
+
+        /**
+         * Checks whether specified {@link PermissionCheck} is implied by this {@link PermissionCheck}.<br>
+         * This is important method to be considered during implementation as it compares {@link PermissionCheck}s.
+         * It is used in scenario's to test if one is granted a permission does that also grant other permission.
+         *
+         * @param permissionCheck {@link PermissionCheck}
+         * @return {@code true} if the specified permission is implied by this {@link PermissionCheck} else
+         * returns {@code false}
+         */
+        boolean implies(PermissionCheck permissionCheck);
+    }
+
+    // Automaton based permission check
+    private static class AutomatonPermissionCheck implements PermissionCheck {
+        private final Automaton automaton;
 
-        public ConditionalClusterPermission(ConfigurableClusterPrivilege conditionalPrivilege) {
-            super(conditionalPrivilege.getPrivilege());
-            this.conditionalPrivilege = conditionalPrivilege;
+        AutomatonPermissionCheck(final Automaton automaton) {
+            this.automaton = automaton;
         }
 
         @Override
-        public boolean check(String action, TransportRequest request) {
-            return super.privilege.predicate().test(action) && conditionalPrivilege.getRequestPredicate().test(request);
+        public boolean check(final String action, final TransportRequest request) {
+            return Automatons.predicate(automaton).test(action);
         }
 
         @Override
-        public List<Tuple<ClusterPrivilege, ConfigurableClusterPrivilege>> privileges() {
-            return Collections.singletonList(new Tuple<>(super.privilege, conditionalPrivilege));
+        public boolean implies(final PermissionCheck permissionCheck) {
+            if (permissionCheck instanceof AutomatonPermissionCheck) {
+                return Operations.subsetOf(((AutomatonPermissionCheck) permissionCheck).automaton, this.automaton);
+            }
+            return false;
```

Comment:
Couldn't we make this work for `ActionRequestPredicatePermissionCheck`, if they have the same action predicate? Or this is out of scope for now?